### ISSUE TITLE
Upgrade to PHPUnit 10

### ISSUE THREAD

**Issue by Wirone** [PHP-CS-Fixer/PHP-CS-Fixer#6884]
## Feature request

We should upgrade to PHPUnit v10, this will also unlock support for [ParaTest](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6883#issuecomment-1497321496).

In order to be able to upgrade PHPUnit to v10 these blockers need to be resolved:
- [x] PHPUnit 10.1 is required (because of [this](https://github.com/sebastianbergmann/phpunit/issues/5201))
- [x] PHPUnitGoodPractices/*
  - ⚠️ ~[BLOCKER](#issuecomment-1675400947)~
  - dropped usage in #7362 & #7363
- [x] phpspec/phpspec#1446 (replacement of ~~phpspec/phpspec#1444~~)
- [x] phpspec/prophecy-phpunit#45
- [ ] phpspec/prophecy#599 (support for `sebastian/recursion-context` v5 required)
- [ ] Allow `phpunit/phpunit` >=10.1 in `composer.json`
- [ ] Update codebase to be compatible with [10.x release](https://github.com/sebastianbergmann/phpunit/blob/main/ChangeLog-10.0.md#1000---2023-02-03), while keeping compatibility with 9.x (v10 requires PHP 8.1 while we support 7.4+)

**Comment by julienfalque** [comment]
If i remember correctly we only need phpspec's packages to use Prophecy and I think we don't use it much. What about replacing those usages with PHPUnit's native mocks and drop the dependencies?

**Comment by Wirone** [comment]
@julienfalque yes, that's the option that would simplify the upgrade process 🙂. I'll look at it and maybe create another PR for that.

**Comment by Wirone** [comment]
@julienfalque I believe we can't get rid of Prophecy just like that because there is whole architecture prepared to ensure `prophesize()` is preferred over mocking ([here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/fb7f3523d563406554a3e739ecc98c25b7943867/tests/TestCase.php#L34) and [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/fb7f3523d563406554a3e739ecc98c25b7943867/tests/TestCase.php#L49-L50)). Yet, it's possible to replace prophesized interfaces with anonymous classes like this:

<img width="1409" alt="image" src="https://user-images.githubusercontent.com/600668/230511908-5ebb215f-c9f0-42ea-933d-29275b29bee8.png">

and test passes without problems. I would prepare PR with it if there were only few usages, but there are over 100 `->prophesize()` in tests, so @keradus has to decide if we want to go this way before investing any time in it.

**Comment by keradus** [comment]
I would not go with native PHPUnit doubles, but keep using Prophecy. Author of PHPUnit too

> If I were to create a new mocking framework today it would probably look a lot like Prophecy
> (~) Sebastian Bergmann


I was also thinking myself that with anonymous classes we can avoid using specially created mocks indeed, but did not saw value in spending time for that transition, myself.

**Comment by Wirone** [comment]
It won't be easy 😩. I was able to change constraints to allow PHPUnit 10 in `phpunitgoodpractices/traits`:

<img width="1206" alt="image" src="https://github.com/PHPUnitGoodPractices/Traits/assets/600668/49d3ede7-fdbc-4070-8d60-476c606c6752">

BUT the whole idea of that repo is killed by `final` added to assert methods of PHPUnit:

```shell
$ vendor/bin/phpunit

Fatal error: Cannot override final method PHPUnit\Framework\Assert::assertEquals() in src/IdentityOverEqualityTrait7.php on line 18
```

Personally I think it would be better if these checks were provided using PHPStan's rules, without the need to hack into test classes' inheritance tree. We could use `spaze/phpstan-disallowed-calls` which has support for [`custom rules`](https://github.com/spaze/phpstan-disallowed-calls/blob/main/docs/custom-rules.md) and could disallow calls like `PHPUnit\Framework\TestCase::getMockBuilder()`. If some current checks can't be migrated using existing rules/plugins, we could write our own. The effect would be the same, but non-invasive like it's now.

@keradus @kubawerlos let me know what you think 🙂.

**Comment by kubawerlos** [comment]
I like that idea. These traits introduce one level of magic and some of the checks they do are already covered by `phpstan/phpstan-strict-rules`.

**Comment by keradus** [comment]
> BUT the whole idea of that repo is killed by final added to assert methods of PHPUnit

traits repo was introduced when some rather not good practices of PHPUnit were allowed and there was no way to prevent them.
That was in time of extremely old PHPUnit and lack of mature SCA around.

if we have good new PHPUnit and SCA tools covering the best practices and recommendations, likely we can drop the `traits` repo

**Comment by Wirone** [comment]
I've spent some time yesterday on it and here's some additional info:

- with `spaze/phpstan-disallowed-calls` it's possible to block `assertEquals()` completely, but it's not possible to do this only for scalar comparisons. So for this particular use case this plugin does not solve our problem
- as @kubawerlos some of the assert cases are handled by [PHPStan's rules](https://github.com/phpstan/phpstan-phpunit), but it does not replace `IdentityOverEqualityTrait` completely
- there's [`ikvasnica/phpstan-clean-test`](https://github.com/ikvasnica/phpstan-clean-test) which provides some interesting rules for more strict test code, but it's not compatible with PHPStan v1 (only `v0.*`, created [issue](https://github.com/ikvasnica/phpstan-clean-test/issues/20))

@keradus do you consider dropping traits without instant replacements? I mean, do we need to port all current checks to new approach to be able to bump PHPUnit?

**Comment by keradus** [comment]
some further steps in #7362 and #7363, bit unblocking us, but some blockers still present

**Comment by keradus** [comment]
@Wirone , we allow PHPUnit v10 and v11. anything else to cover here ?

**Comment by Wirone** [comment]
@keradus see last point on the checklist - we need to get rid of PHP 7.4 if I remember correctly, or customise CI for 7.4 because most probably ParaUnit won't work with it.

**Comment by keradus** [comment]
ticket purpose: upgrade to PHPUnit 10
current status: we use PHPUnit 10 and 11

from description: let's use ParaUnit
from discussion: we can use any, ParaUnit or ParaTest
current status: we use ParaTest already


anything open to achieve in this ticket?

**Comment by keradus** [comment]
closing as imho already done year ago ;)
